### DISCUSSION TITLE
-fhosted and -ffreestanding Compiles Flags

### DISCUSSION THREAD

**Discussion by marcrittinghaus** [unikraft/unikraft#668] [General]
While building a cpp application, I noticed that I always get the warning that `fhosted` is not supported for C++. 

![image](https://user-images.githubusercontent.com/8928900/204286744-611d8acc-6628-4995-b666-f887c4c7d423.png)

I checked where we set the parameter and found the following in [`unikraft/Makefile.uk:94`](https://github.com/unikraft/unikraft/blob/413b1f06d2d95eae6c7526f4f1c793c2c90066a9/Makefile.uk#L94)
```C
COMPFLAGS-y += -fhosted -ffreestanding -fno-tree-loop-distribute-patterns
```
From my understanding `-fhosted`, which is the default, means that the compiler can expect the standard library to be present and have the standard-conforming implementation. So, it is possible for the compiler to optimize a `memset` by replacing it with inline ASM instructions.

An application like this:
```C
char buf[1024];

int main(int argc, char *argv[])
{
        memset(buf, 0xff, sizeof(buf));
        return 0;
}
```
could be compiled to this with `-O3`:
```ASM
  10d170:       ba 80 2d 19 00          mov    edx,0x192d80               # Address of buf
  10d175:       48 c7 c0 ff ff ff ff    mov    rax,0xffffffffffffffff     # Value to set
  10d17c:       b9 80 00 00 00          mov    ecx,0x80                   # Length to set
  10d181:       48 89 d7                mov    rdi,rdx
  10d184:       f3 48 ab                rep stos QWORD PTR es:[rdi],rax   # Perform memset
  10d187:       31 c0                   xor    eax,eax                    # retval = 0
  10d189:       c3                      ret
```
However, currently for Unikraft this will compile to:
```ASM
  10d110:       55                      push   rbp
  10d111:       ba 00 04 00 00          mov    edx,0x400                  # Length in bytes
  10d116:       be ff 00 00 00          mov    esi,0xff                   # Value to set
  10d11b:       bf 80 3d 19 00          mov    edi,0x193d80               # Address of buf
  10d120:       31 c0                   xor    eax,eax
  10d122:       48 89 e5                mov    rbp,rsp
  10d125:       e8 26 53 06 00          call   172450 <memset>            # Perform memset
  10d12a:       31 c0                   xor    eax,eax
  10d12c:       5d                      pop    rbp
  10d12d:       c3                      ret
```
This is due to the `-ffreestanding` parameter that will let the compiler assume that there is no standard library and thus functions like `memset` are more or less regular functions. Optimizing them is obviously not possible in this case.

The line is introduced by commit [86fef08ac53](https://github.com/unikraft/unikraft/commit/86fef08ac53b06ce48a4160f8569a3c8d2c2b65c), which states it introduces the parameters to avoid a recursive call bug for GCC 10.

I was wondering, if (1) the consequences of this where clear, and (2) why both parameters where specified because `-ffreestanding` takes precedence over `-fhosted`, thus potentially preventing a lot of optimizations. @vladandrew You were reviewing this. Can you remember any details?

After discussing the topic also with @mschlumpp, my suggestion is to use `-ffreestanding` only for `nolibc` to be safe with non-standard behavior but remove it for musl and newlib. Also `-fhosted` seems redundant as it is the default for C (if not freestanding) and causes warnings for cpp files.

For a more thorough discussion of hosted vs. freestanding also look [here](https://stackoverflow.com/questions/30825151/is-there-a-meaningful-distinction-between-freestanding-and-hosted-implementation).

**Comment by vladandrew** [comment]
Hi Marc,

I remember we had to do this for a quick fix during the AE process for EuroSys, it was related to a bug in DPDK. I agree with your research, we should use `-freestanding` only for `nolibc`.

**Comment by razvand** [comment]
This is now discussed in #740 , let's move the discussion there.
